[SPIRV] add pre legalization instruction combine

farzonl opened the pull request:
- Add the boilerplate to support instcombine in SPIRV
- instcombine length(X-Y) to distance(X,Y)
- switch HLSL's distance intrinsic to not special case for SPIRV.
- fixes #122766
- This RFC we were requested to add in the infra for pattern matching: https://discourse.llvm.org/t/rfc-add-targetbuiltins-for-spirv-to-support-hlsl/83329/13

farzonl committed:
[SPIRV] add pre legalization instruction combine
- Add the boilerplate to support instcombine in SPIRV
- instcombine length(X-Y) to distance(X,Y)
- switch HLSL's distance intrinsic to not special case for SPIRV.
- fixes #122766

farzonl commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/distance.ll:
```diff
@@ -1,33 +1,44 @@
+  ; CHECK: %[[#arg1:]] = OpFunctionParameter %[[#vec4_float_32]]
+  ; CHECK: %[[#]] = OpExtInst %[[#float_32]] %[[#op_ext_glsl]] Distance %[[#arg0]] %[[#arg1]]
+  %delta = fsub  <4 x float> %a, %b
+  %spv.length = call float @llvm.spv.length.f32(<4 x float> %delta)
```
this is the test case to test the instcombine.

farzonl commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/distance.ll:
```diff
@@ -1,33 +1,44 @@
-; RUN: llc -O0 -mtriple=spirv-unknown-unknown %s -o - | FileCheck %s
```
changed line endings from crlf to lf.

michalpaszkowski commented on llvm/lib/Target/SPIRV/SPIRVCombine.td:
```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
```
Should the file include some comment? Otherwise, I think this part could be removed.

farzonl committed:
address pr comments

VyacheslavLevytskyy commented:
@farzonl Thank you for adding instcombine/pattern matching capabilities. LGTM overall.
Maybe we should also document each transformation a bit in comments to the corresponding function, so that we clearly see what kind of input we expect to see to run the substitution? It could be anything in style of lib/Transforms/InstCombine/InstructionCombining.cpp comments (e.g., `(op (cast (op X, C2)), C1) --> (cast (op X, op (C1, C2)))`) or in style of more lengthy comments in lib/Target/SPIRV/SPIRVPreLegalizer.cpp (e.g., `propagateSPIRVType`). What do you think?

farzonl commented:
> I see also some problems in test cases

@VyacheslavLevytskyy the SPIRV test case failures are not related to my change. They are also failing on main.

VyacheslavLevytskyy commented:
> > I see also some problems in test cases
> 
> @VyacheslavLevytskyy the SPIRV test case failures are not related to my change. They are also failing on main.

I will check now. For one the reason is clear, https://github.com/llvm/llvm-project/pull/123191

VyacheslavLevytskyy commented:
> > > I see also some problems in test cases
> > 
> > 
> > @VyacheslavLevytskyy the SPIRV test case failures are not related to my change. They are also failing on main.
> 
> I will check now. For one the reason is clear, #123191

I can't reproduce two other fails on different environments, including actually Github Actions, so it maybe that after you rebase to `main` we will see all green now.

farzonl committed:
address pr comments describing transformation

farzonl commented:
> > > > I see also some problems in test cases
> > > 
> > > 
> > > @VyacheslavLevytskyy the SPIRV test case failures are not related to my change. They are also failing on main.
> > 
> > 
> > I will check now. For one the reason is clear, #123191
> 
> I can't reproduce two other fails on different environments, including actually Github Actions, so it maybe that after you rebase to `main` we will see all green now.

I rebased and the tests are passing locally. but they are still failing in ci.
```bash
bash-3.2$ ../llvm_debug_build/bin/llvm-lit -sv llvm/test/CodeGen/SPIRV/branching/OpSwitchUnreachable.ll

Testing Time: 2.29s

Total Discovered Tests: 1
  Passed: 1 (100.00%)
bash-3.2$ ../llvm_debug_build/bin/llvm-lit -sv llvm/test/CodeGen/SPIRV/keep-tracked-const.ll

Testing Time: 0.21s

Total Discovered Tests: 1
  Passed: 1 (100.00%)
```

VyacheslavLevytskyy commented:
> I rebased and the tests are passing locally. but they are still failing in ci.

It's the strange part. The thing is when I created a new https://github.com/llvm/llvm-project/pull/123191 today, I see all tests passing in CI before I press Merged. That's why I thought that we can see the same here after rebase. This https://github.com/llvm/llvm-project/pull/123191 today was created basing on fresh main. So, it's a bit weird.

farzonl committed:
- Add llc-pipeline test to track pass run ordering
- move instcombine before spirv-prelegalizer  simplifies the matcher
- update matcher and inst combine application
- make SPIRVPreLegalizer preserve previous passes.

farzonl commented on llvm/lib/Target/SPIRV/SPIRVPreLegalizerCombiner.cpp:
```diff
@@ -0,0 +1,260 @@
+  bool EnableOpt =
+      MF.getTarget().getOptLevel() != CodeGenOptLevel::None && !skipFunction(F);
+  GISelKnownBits *KB = &getAnalysis<GISelKnownBitsAnalysis>().get(MF);
+  MachineDominatorTree *MDT =
```
Here is something I would like some feedback on. My pass doesn't take advantage of CSE or DomTrees. The way that `*PreLegalizerCombinerImpl` in our case `SPIRVPreLegalizerCombinerImpl` works is it needs a `MDT` and a `CSEInfo`.  However, If I do what the MIPS backend does in `llvm/lib/Target/Mips/MipsPreLegalizerCombiner.cpp` I can set these to nullptr and have less required passes that need to be preserved. This is however not how RISCV and the AArch64 backends are doing their combiner passes. Does it make more sense to follow the MIPS pattern and eventually evolve to become like the RISCV/AArch64 pattern when we have a need for the extra passes or to start out where RISCV and AArch64 are at?

VyacheslavLevytskyy commented on llvm/test/CodeGen/SPIRV/llc-pipeline.ll:
```diff
@@ -0,0 +1,140 @@
+; RUN: llc -mtriple=spirv-unknown-unknown -debug-pass=Structure < %s -o /dev/null 2>&1 | \
```
Just as a thought aloud, I wonder how stable this list can be at this stage of SPIR-V Backend development? I guess this test case may be subjected to more frequent changes than others.
Also I think we'd anyway need to fix -O0 or -Osomething here.

farzonl replied:
I'm using this test case to help me figure out why I'm seeing an assert fire in `SPV_INTEL_inline_assembly/inline_asm.ll`. I can remove it before I take this PR out of draft. TLDR on the investigation there is an assert firing in `llvm/lib/Target/SPIRV/SPIRVPreLegalizer.cpp` The pass before mine had an inline assembly that is  some how removed by my pass.

pass before:
```bash
/mnt/DevDrive/projects/llvm_debug_build/bin/llc -stop-after=irtranslator -global-isel \
 -verify-machineinstrs -O0 -mtriple=spirv64-unknown-unknown \
llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_inline_assembly/inline_asm.ll  \
--spirv-ext=+SPV_INTEL_inline_assembly -o -
```
Inline assembly removed:
```asm
%71:_(s64) = G_ANYEXT %68(s32)
%70:iid = COPY %71(s64)
%72:fid = COPY %69(s64)
INLINEASM &"constcmd $0 $1", 0 /* attdialect */, 327689 /* reguse:iID */, %70, 262153 /* reguse:fID */, %72
```

pass after:
```
/mnt/DevDrive/projects/llvm_debug_build/bin/llc -stop-after=spirv-prelegalizer-combiner -global-isel \
 -verify-machineinstrs -O0 -mtriple=spirv64-unknown-unknown \
llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_inline_assembly/inline_asm.ll  \
--spirv-ext=+SPV_INTEL_inline_assembly -o -
```
MIR diff
![image](https://github.com/user-attachments/assets/7feb8896-effc-4ce8-9787-6ccac09e96cf)

I've tried removing the CSE analysis pass but something else in `SPIRVPreLegalizerCombiner` is removing the inline assembly.
![image](https://github.com/user-attachments/assets/a48b55b5-775f-45d0-8cc4-c02589ca12d7)

s-perron reviewed (approved):
I don't see any problem, but I'm not familiar with standard practices in llvm. I like having the combiner pass.

farzonl committed:
remove CSE pass dependency

farzonl committed:
turn off DCE

VyacheslavLevytskyy commented:
@farzonl Was enabled GISelCSEAnalysisWrapperPass a reason of problems with inline_asm?

farzonl commented:
> @farzonl Was enabled GISelCSEAnalysisWrapperPass a reason of problems with inline_asm?

It was CSE and The combiner pass does dead code elimination. I removed one and turned the other off. Their is a bug in the SPIRVPreLegalizer where its expecting a matching number of inline asm and the spv inline asm intrinsics. Im not sure what the fix for the bug should be so it was simpler to just turn off cse and dce.

VyacheslavLevytskyy commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/distance.ll:
```diff
@@ -1,33 +1,44 @@
-
-declare half @llvm.spv.distance.f16(<4 x half>, <4 x half>)
-declare float @llvm.spv.distance.f32(<4 x float>, <4 x float>)
+; RUN: llc -O0 -mtriple=spirv-unknown-unknown %s -o - | FileCheck %s
```
Can we please add -verify-machineinstrs here to be sure that expensive_checks run will be happy?

farzonl committed:
address pr comments add -verify-machineinstrs to distance.ll
Also add a MIR test to test the instcombine independent of other passes.
